frontend/src/pages/Queue.tsx

```diff
@@ -0,0 +1,100 @@
+import LocationSelection from "./LocationSelection/LocationSelection";
+import { useState, useEffect, useRef } from "react";
+import { getExpectedWaitTime } from '../utils/api';
+import { useNavigate } from 'react-router-dom';
+
+const Queue = () => {
+  const [waitTime, setWaitTime] = useState<number | null>(null);
+  const [isQueueFull, setIsQueueFull] = useState<boolean>(false);
+  const selectedLocation = localStorage.getItem('selectedLocation');
+  const navigate = useNavigate();
+  const cacheExpiryTimeoutRef = useRef<ReturnType<typeof setTimeout> | null>(null);
+  const CACHE_EXPIRY_THRESHOLD = 60 * 1000;
+
+  useEffect(() => {
+    // Function to fetch wait time and manage cache
+    const fetchWaitTime = async () => {
+      const cachedWaitTime = localStorage.getItem("expectedWaitTime");
+      const lastCheckTime = localStorage.getItem("expectedWaitTimeLastCheckTime");
+      const currentTime = Date.now();
+      const cacheAge = lastCheckTime ? currentTime - parseInt(lastCheckTime) : null;
+
+      // Use cached data if it's valid
+      if (cachedWaitTime && cacheAge !== null && cacheAge < CACHE_EXPIRY_THRESHOLD) {
+        const cachedTime = parseFloat(cachedWaitTime);
+        setWaitTime(cachedTime);
+        setIsQueueFull(cachedTime >= 5);
+
+        // Schedule cache expiry if the user remains on the page
+        const timeTillCacheExpiry = CACHE_EXPIRY_THRESHOLD - cacheAge + 10;
+        cacheExpiryTimeoutRef.current = setTimeout(() => {
+          localStorage.removeItem("expectedWaitTime");
+          localStorage.removeItem("expectedWaitTimeLastCheckTime");
+        }, timeTillCacheExpiry);
+      } else {
+        try {
+          // Fetch updated data from API
+          const data = await getExpectedWaitTime(selectedLocation);
+          const expectedWaitTime = data.expectedWaitTime;
+
+          setWaitTime(expectedWaitTime);
+          setIsQueueFull(expectedWaitTime >= 5);
+
+          // Cache the updated data
+          localStorage.setItem("expectedWaitTime", expectedWaitTime.toString());
+          localStorage.setItem("expectedWaitTimeLastCheckTime", currentTime.toString());
+
+          // Schedule cache expiry
+          cacheExpiryTimeoutRef.current = setTimeout(() => {
+            localStorage.removeItem("expectedWaitTime");
+            localStorage.removeItem("expectedWaitTimeLastCheckTime");
+          }, CACHE_EXPIRY_THRESHOLD);
+        } catch (error) {
+          console.error("Failed to fetch expected wait time:", error);
+        }
+      }
+    };
+
+    // Fetch wait time if a location is selected
+    if (selectedLocation) fetchWaitTime();
+
+    // Cleanup function to clear timeout on unmount
+    return () => {
+      if (cacheExpiryTimeoutRef.current) {
+        clearTimeout(cacheExpiryTimeoutRef.current);
+        cacheExpiryTimeoutRef.current = null;
+      }
+    };
+  }, [selectedLocation]);
```

Comment:
Use visibility API to avoid having the timer run and spend reads to Firestore when the user is not actively on the page. Refer to current state as an example

Reply:
Please add screenshots of the console logs showing your timers starting on and off. For example, [PR 42](https://github.com/uoftblueprint/brampton-tennis-queue/pull/42)